core/services/sleeper_task.go

```diff
@@ -32,27 +51,68 @@ func NewSleeperTask(worker Worker) SleeperTask {
 		worker: worker,
 		waker:  make(chan struct{}, 1),
 		closer: make(chan struct{}, 1),
+		closed: make(chan struct{}, 1),
+		state:  unstarted,
+		mutex:  sync.Mutex{},
 	}
 }
 
 // Start begins the SleeperTask
+// Calling start on a started task will return an error
 func (s *sleeperTask) Start() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	switch s.state {
+	case stopped:
+		return errors.New("cannot start a sleeper task that has been stopped")
+	case started:
+		return errors.New("sleeper task is already started")
+	}
+	s.state = started
 	go s.workerLoop()
 	return nil
 }
 
 // Stop stops the SleeperTask
+// It can only be called once
+// It will block until all tasks have completed, or the timeout is reached
+// Stopped tasks cannot be restarted
 func (s *sleeperTask) Stop() error {
+	s.mutex.Lock()
+	if s.state == stopped {
+		s.mutex.Unlock()
+		return errors.New("sleeper task is already stopped")
+	}
+	s.state = stopped
+	s.mutex.Unlock()
+
 	s.closer <- struct{}{}
-	return nil
+	// NOTE: Closing the channels will cause the rogue task to panic if it does eventually complete
+	defer close(s.waker)
+	defer close(s.closer)
+	defer close(s.closed)
+	select {
+	case <-s.closed:
+		return nil
+	case <-time.After(stopWaitTime):
+		return fmt.Errorf("task did not complete within %s", stopWaitTime.String())
+	}
 }
 
 // WakeUp wakes up the sleeper task, asking it to execute its Worker.
-func (s *sleeperTask) WakeUp() {
+// Idempotent, can be called multiple times but will only wake the worker once (until the worker finishes again)
+func (s *sleeperTask) WakeUp() error {
```

Comment:
@j16r says: 

I would take out the WakeUp error

just panic

an error at that point is not recoverable